### PR TITLE
Use new ClonePath peramater from src-cli

### PR DESCRIPTION
**Pull request by varsanojidan**
Reopening [this](https://github.com/sourcegraph/sourcegraph/pull/40170) in a new PR

Directly related to https://github.com/sourcegraph/src-cli/pull/810.

## Test plan
Tested locally using src-cli serving regular git and bare git repos.

### CODE REVIEW

File: internal/repos/other.go [hunk 1]
```diff
@@ -182,42 +191,43 @@ func (s OtherSource) srcExpose(ctx context.Context) ([]*types.Repo, error) {
...
+	repos := make([]*types.Repo, 0, len(data.Items))
 	for _, r := range data.Items {
-		// The only required field is URI
-		if r.URI == "" {
-			return nil, errors.Errorf("repo without URI returned from src-expose: %+v", r)
+		repo := &types.Repo{
+			URI: r.URI,
+		}
+		// The only required fields are URI and ClonePath
+		if r.URI == "" || r.ClonePath == "" {
```
Comment (varsanojidan) [inline internal/repos/other.go:200]: I was debating about whether it made sense to return an error here if they are using an outdated src-cli in order to force them to upgrade the src-cli version, or to add a failsafe that would work with the old src-clis but would add business logic and code we need to remove later.

Comment (keegancsmith) [reply internal/repos/other.go:200]: Personally I'd treat this how we normally deprecate an API. IE log a warning for a release or two then hard fail.

If you do want to go this route, please make the error more helpful so the admin can quickly resolve the problem. IE tell them they need to upgrade src-cli and make this clear in the changelog.

File: internal/repos/other.go [hunk 2]
```diff
@@ -169,7 +178,7 @@ func (s OtherSource) srcExpose(ctx context.Context) ([]*types.Repo, error) {
 	}
 
 	var data struct {
-		Items []*types.Repo
+		Items []*srcExposeItem
```
Comment (keegancsmith) [inline internal/repos/other.go:181]: FYI This was never used, but I intentionally made this a types.Repo so in future other APIs could influence fields like star count. You might be able to keep that behaviour if you embed types.Repo into srcExposeItem, but probably better to limit the scope and do it this way :)

**Review by keegancsmith — APPROVED** [review]

File: internal/repos/other.go (hunk 1 above)
Comment (ryanslade) [reply internal/repos/other.go:200]: Agreed that we should try and be backwards compatible if it doesn't add too much complexity to the code.

Comment (varsanojidan) [reply internal/repos/other.go]: Sounds good, I've added a deprecation warning

File: internal/repos/other.go [hunk 3]
```diff
@@ -8,6 +8,7 @@ import (
 	"net/url"
 	"strings"
 
+	"github.com/inconshreveable/log15"
```
Comment (keegancsmith) [inline internal/repos/other.go:11]: nit: don't introduce uses of log15, use sourcegraph/log.

File: internal/repos/other.go [hunk 4]
```diff
@@ -182,42 +192,54 @@ func (s OtherSource) srcExpose(ctx context.Context) ([]*types.Repo, error) {
...
+	loggedDeprecationError := false
 	for _, r := range data.Items {
-		// The only required field is URI
+		repo := &types.Repo{
+			URI: r.URI,
+		}
+		// The only required fields are URI and ClonePath
 		if r.URI == "" {
-			return nil, errors.Errorf("repo without URI returned from src-expose: %+v", r)
+			return nil, errors.Errorf("repo without URI and/or ClonePath returned from src-expose: %+v", r)
```
Comment (keegancsmith) [inline internal/repos/other.go:203]: error is not correct until you enforce it. ie undo this hunk

File: internal/repos/other.go [hunk 5]
```diff
@@ -182,42 +192,54 @@ func (s OtherSource) srcExpose(ctx context.Context) ([]*types.Repo, error) {
...
+		}
+
+		// ClonePath is always set in the new versions of src-cli.
+		// TODO: @varsanojidan Remove this by version 3.45.0 and add it to the check above.
+		if r.ClonePath == "" {
+			if !loggedDeprecationError {
+				log15.Warn("The version of src-cli serving git repositories is deprecated, please upgrade to the latest version.")
+				loggedDeprecationError = true
+			}
+			r.ClonePath = r.URI + "/.git"
```
Comment (keegancsmith) [inline internal/repos/other.go:213]: you need to check that it doesn't have the `.git` suffix first.

**Pull request merged**